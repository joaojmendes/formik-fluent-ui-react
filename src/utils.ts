import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
