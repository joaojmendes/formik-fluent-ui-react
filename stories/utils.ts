import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
