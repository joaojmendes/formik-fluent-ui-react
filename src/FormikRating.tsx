import { IRatingProps, Rating } from '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToRating<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<IRatingProps, 'value' | 'onChange'> {
  return {
    value: field.value,
    onChange: (_, value) => {
      form.setFieldValue(field.name, value)
      field.onBlur(createFakeEvent(field))
    },
  }
}

export type FormikRatingProps<T = any> = Omit<
  IRatingProps,
  'value' | 'onChange' | 'onBlur' | 'form'
> &
  FieldProps<T>

export function FormikRating<T = any>({
  field,
  form,
  ...props
}: FormikRatingProps<T>) {
  return <Rating {...props} {...mapFieldToRating({ field, form })} />
}
