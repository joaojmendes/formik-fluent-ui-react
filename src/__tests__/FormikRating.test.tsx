import { Field, FieldProps, Form, Formik } from 'formik'
import { Rating, setIconOptions } from '@fluentui/react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikRating, mapFieldToRating } from '../FormikRating'
import { noop, serialize } from './utils'

// Suppress icon warnings.
setIconOptions({
  disableWarnings: true,
})

class Values {
  public rating: number = 3
}

function createFieldProps(value: number = 3): FieldProps<number> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'rating',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
    meta: {}
  } as any
}

test('<FormikRating /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="rating" label="Rating" component={FormikRating} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikRating /> renders a Fabric <Rating />', () => {
  const fieldProps = createFieldProps()

  const formikRating = renderer.create(<FormikRating {...fieldProps} />)
  const fabricRating = renderer.create(
    <Rating {...mapFieldToRating(fieldProps)} />
  )
  expect(serialize(formikRating)).toBe(serialize(fabricRating))
})

test('mapFieldToRating() maps FieldProps to IRatingProps', () => {
  const { field, form, meta } = createFieldProps()
  const props = mapFieldToRating({ form, field, meta })

  expect(props.rating).toBe(field.value)

  props.onChange!(null as any, 5)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, 5)

  // onChange also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
