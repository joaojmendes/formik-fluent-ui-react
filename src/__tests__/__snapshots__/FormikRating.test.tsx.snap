// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikRating /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    aria-label=""
    className="ms-Rating-star ms-RatingStar-root root-47 ms-RatingStar-root--small rootIsSmall-48"
    id="Rating1"
  >
    <div
      className="ms-FocusZone ms-Rating-focuszone ratingFocusZone-57 ms-RatingStar-root--small rootIsSmall-48"
      data-focuszone-id="FocusZone3"
      data-is-focusable={false}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDownCapture={[Function]}
      role="presentation"
      tabIndex={-1}
    >
      <button
        className="ms-Rating-button ratingButton-53 ms-Rating--small ratingStarIsSmall-54"
        data-is-current={true}
        disabled={false}
        id="Rating1-star-0"
        onClick={[Function]}
        onFocus={[Function]}
        role="presentation"
        type="button"
      >
        <span
          className="ms-Rating-labelText labelText-56"
          id="RatingLabel2-1"
        >
          
        </span>
        <div
          className="ms-RatingStar-container ratingStar-50"
        >
          <i
            aria-hidden={true}
            className="ms-RatingStar-back ratingStarBack-58"
            data-icon-name="FavoriteStarFill"
            role="presentation"
          />
          <i
            aria-hidden={true}
            className="ms-RatingStar-front ratingStarFront-59"
            data-icon-name="FavoriteStarFill"
            role="presentation"
            style={
              Object {
                "width": "100%",
              }
            }
          />
        </div>
      </button>
      <button
        className="ms-Rating-button ratingButton-53 ms-Rating--small ratingStarIsSmall-54"
        disabled={false}
        id="Rating1-star-1"
        onClick={[Function]}
        onFocus={[Function]}
        role="presentation"
        type="button"
      >
        <span
          className="ms-Rating-labelText labelText-56"
          id="RatingLabel2-2"
        >
          
        </span>
        <div
          className="ms-RatingStar-container ratingStar-50"
        >
          <i
            aria-hidden={true}
            className="ms-RatingStar-back ratingStarBack-58"
            data-icon-name="FavoriteStarFill"
            role="presentation"
          />
          <i
            aria-hidden={true}
            className="ms-RatingStar-front ratingStarFront-59"
            data-icon-name="FavoriteStarFill"
            role="presentation"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
      </button>
      <button
        className="ms-Rating-button ratingButton-53 ms-Rating--small ratingStarIsSmall-54"
        disabled={false}
        id="Rating1-star-2"
        onClick={[Function]}
        onFocus={[Function]}
        role="presentation"
        type="button"
      >
        <span
          className="ms-Rating-labelText labelText-56"
          id="RatingLabel2-3"
        >
          
        </span>
        <div
          className="ms-RatingStar-container ratingStar-50"
        >
          <i
            aria-hidden={true}
            className="ms-RatingStar-back ratingStarBack-58"
            data-icon-name="FavoriteStarFill"
            role="presentation"
          />
          <i
            aria-hidden={true}
            className="ms-RatingStar-front ratingStarFront-59"
            data-icon-name="FavoriteStarFill"
            role="presentation"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
      </button>
      <button
        className="ms-Rating-button ratingButton-53 ms-Rating--small ratingStarIsSmall-54"
        disabled={false}
        id="Rating1-star-3"
        onClick={[Function]}
        onFocus={[Function]}
        role="presentation"
        type="button"
      >
        <span
          className="ms-Rating-labelText labelText-56"
          id="RatingLabel2-4"
        >
          
        </span>
        <div
          className="ms-RatingStar-container ratingStar-50"
        >
          <i
            aria-hidden={true}
            className="ms-RatingStar-back ratingStarBack-58"
            data-icon-name="FavoriteStarFill"
            role="presentation"
          />
          <i
            aria-hidden={true}
            className="ms-RatingStar-front ratingStarFront-59"
            data-icon-name="FavoriteStarFill"
            role="presentation"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
      </button>
      <button
        className="ms-Rating-button ratingButton-53 ms-Rating--small ratingStarIsSmall-54"
        disabled={false}
        id="Rating1-star-4"
        onClick={[Function]}
        onFocus={[Function]}
        role="presentation"
        type="button"
      >
        <span
          className="ms-Rating-labelText labelText-56"
          id="RatingLabel2-5"
        >
          
        </span>
        <div
          className="ms-RatingStar-container ratingStar-50"
        >
          <i
            aria-hidden={true}
            className="ms-RatingStar-back ratingStarBack-58"
            data-icon-name="FavoriteStarFill"
            role="presentation"
          />
          <i
            aria-hidden={true}
            className="ms-RatingStar-front ratingStarFront-59"
            data-icon-name="FavoriteStarFill"
            role="presentation"
            style={
              Object {
                "width": "0%",
              }
            }
          />
        </div>
      </button>
    </div>
  </div>
</form>
`;
